File: Server/src/main/core/cache/crypto/XTEACryption.java
package core.cache.crypto;

import java.nio.ByteBuffer;

public final class XTEACryption {

    private static final int DELTA = -1640531527;

    private static final int SUM = -957401312;

    private static final int NUM_ROUNDS = 32;

    private XTEACryption() {
        /*
         * empty.
         */
    }

    public static ByteBuffer decrypt(int[] keys, ByteBuffer buffer) {
        return decrypt(keys, buffer, buffer.position(), buffer.limit());
    }

    public static ByteBuffer decrypt(int[] keys, ByteBuffer buffer, int offset, int length) {
        int numBlocks = (length - offset) / 8;
        int[] block = new int[2];
        for (int i = 0; i < numBlocks; i++) {
            int index = i * 8 + offset;
            block[0] = buffer.getInt(index);
            block[1] = buffer.getInt(index + 4);
            decipher(keys, block);
            buffer.putInt(index, block[0]);
            buffer.putInt(index + 4, block[1]);
        }
        return buffer;
    }

    private static void decipher(int[] keys, int[] block) {
        long sum = SUM;
        for (int i = 0; i < NUM_ROUNDS; i++) {
            block[1] -= (int) (keys[(int) ((sum & 0x1933) >>> 11)] + sum ^ block[0] + (block[0] << 4 ^ block[0] >>> 5));
            sum -= DELTA;
            block[0] -= (int) ((block[1] << 4 ^ block[1] >>> 5) + block[1] ^ keys[(int) (sum & 0x3)] + sum);
        }
    }

    public static void encrypt(int[] keys, ByteBuffer buffer) {
        encrypt(keys, buffer, buffer.position(), buffer.limit());
    }

    public static void encrypt(int[] keys, ByteBuffer buffer, int offset, int length) {
        int numBlocks = (length - offset) / 8;
        int[] block = new int[2];
        for (int i = 0; i < numBlocks; i++) {
            int index = i * 8 + offset;
            block[0] = buffer.getInt(index);
            block[1] = buffer.getInt(index + 4);
            encipher(keys, block);
            buffer.putInt(index, block[0]);
            buffer.putInt(index + 4, block[1]);
        }
    }

    private static void encipher(int[] keys, int[] block) {
        long sum = 0;
        for (int i = 0; i < NUM_ROUNDS; i++) {
            block[0] += (int) ((block[1] << 4 ^ block[1] >>> 5) + block[1] ^ keys[(int) (sum & 0x3)] + sum);
            sum += DELTA;
            block[1] += (int) (keys[(int) ((sum & 0x1933) >>> 11)] + sum ^ block[0] + (block[0] << 4 ^ block[0] >>> 5));
        }
    }
}

File: Emulator/src/main/core/cache/crypto/XTEACryption.kt
package core.cache.crypto

import java.nio.ByteBuffer

/**
 * Holds XTEA cryption methods.
 * @author Emperor
 */
object XTEACryption {
    /**
     * The delta value
     */
    private const val DELTA = -1640531527

    /**
     * The sum.
     */
    private const val SUM = -957401312

    /**
     * The amount of "cryption cycles".
     */
    private const val NUM_ROUNDS = 32
    /**
     * Decrypts the buffer data.
     *
     * @param keys   The keys.
     * @param buffer The buffer to decrypt.
     * @param offset The offset of the data to decrypt.
     * @param length The length.
     * @return The decrypted data.
     */
    /**
     * Decrypts the contents of the buffer.
     *
     * @param keys   The cryption keys.
     * @param buffer The buffer.
     * @return the byte buffer
     */
    @JvmStatic
    @JvmOverloads
    fun decrypt(
        keys: IntArray,
        buffer: ByteBuffer,
        offset: Int = buffer.position(),
        length: Int = buffer.limit()
    ): ByteBuffer {
        val numBlocks = (length - offset) / 8
        val block = IntArray(2)
        for (i in 0 until numBlocks) {
            val index = i * 8 + offset
            block[0] = buffer.getInt(index)
            block[1] = buffer.getInt(index + 4)
            decipher(keys, block)
            buffer.putInt(index, block[0])
            buffer.putInt(index + 4, block[1])
        }
        return buffer
    }

    /**
     * Deciphers the values.
     *
     * @param keys  The cryption key.
     * @param block The values to decipher.
     */
    private fun decipher(keys: IntArray, block: IntArray) {
        var sum = SUM.toLong()
        repeat(NUM_ROUNDS) {
            block[1] -= (keys[(sum and 0x1933L ushr 11).toInt()] + sum xor (block[0] + (block[0] shl 4 xor (block[0] ushr 5))).toLong()).toInt()
            sum -= DELTA.toLong()
            block[0] -= (((block[1] shl 4 xor (block[1] ushr 5)) + block[1]).toLong() xor keys[(sum and 0x3L).toInt()] + sum).toInt()
        }
    }
    /**
     * Encrypts the buffer data.
     *
     * @param keys   The keys.
     * @param buffer The buffer to encrypt.
     * @param offset The offset of the data to encrypt.
     * @param length The length.
     * @return The encrypted data.
     */
    /**
     * Encrypts the contents of the byte buffer.
     *
     * @param keys   The cryption keys.
     * @param buffer The buffer to encrypt.
     */
    @JvmOverloads
    fun encrypt(keys: IntArray, buffer: ByteBuffer, offset: Int = buffer.position(), length: Int = buffer.limit()) {
        val numBlocks = (length - offset) / 8
        val block = IntArray(2)
        for (i in 0 until numBlocks) {
            val index = i * 8 + offset
            block[0] = buffer.getInt(index)
            block[1] = buffer.getInt(index + 4)
            encipher(keys, block)
            buffer.putInt(index, block[0])
            buffer.putInt(index + 4, block[1])
        }
    }

    /**
     * Enciphers the values of the block.
     *
     * @param keys  The cryption keys.
     * @param block The block to encipher.
     */
    private fun encipher(keys: IntArray, block: IntArray) {
        var sum: Long = 0
        repeat(NUM_ROUNDS) {
            block[0] += (((block[1] shl 4 xor (block[1] ushr 5)) + block[1]).toLong() xor keys[(sum and 0x3L).toInt()] + sum).toInt()
            sum += DELTA.toLong()
            block[1] += (keys[(sum and 0x1933L ushr 11).toInt()] + sum xor (block[0] + (block[0] shl 4 xor (block[0] ushr 5))).toLong()).toInt()
        }
    }
}